### ISSUE TITLE
Improve CSR Checks when integrating with external Certificate Authorities.

### ISSUE THREAD

**Issue by shankgan** [istio/istio#29838]
When functioning as an RA to validate CSR Requests and forward them to an external CA, Istiod needs to perform more validations on the  x509 CSR Request. 

Note that this code path (where Istiod functions as an RA) is only activated for experimental feature use https://istio.io/latest/docs/tasks/security/cert-management/custom-ca-k8s/.

Some missing validation checks include:
(a) keyUsage and extKeyUsage checks -> This could allow an attacker to request a CA certificate, request certs for code signing etcetera. 
(b) Validating the CN -> This could allow an attacker to request an arbitrary CN in the certificate. 

This Feature request is security related.

**Comment by pmansour** [comment]
One other important field to validate is the `basicConstraints` extension ([RFC 5280 section 4.2.1.9](https://tools.ietf.org/html/rfc5280#section-4.2.1.9)), which defines whether or not the request is for a CA certificate, and the maximum chain depth of any other CAs under this one. When performing RA checks on behalf of external Certificate Authorities, it might make sense to always require this field to be omitted, or if present, for its `cA` boolean to be `false`.

There may also be other relevant fields to validate here, but the ones mentioned so far (KU, EKU, Subject CN and basicConstraints) are the main ones that come to mind right now. Better yet - if there were some way to limit the set of allowed X.509 fields and extensions by default to the ones known and validated by Istio, that would probably result in better security posture.

**Comment by pmansour** [comment]
**Update** - after discussion with other X.509 experts, I think it would be strongly preferable to implement an "allowlist" approach that limits the set of X.509 extensions that can be set on the CSR, to avoid having unvalidated properties blindly copied over by the CA.

Problems with a "denylist" approach (where specific extensions are disallowed but everything else is allowed) can be divided into two categories:

1. **Dangerous general-purpose X.509 extensions**: Some extensions (such as [CRLDistributionPoints](https://tools.ietf.org/html/rfc5280#section-4.2.1.13), [AuthorityInformationAccess](https://tools.ietf.org/html/rfc5280#section-4.2.2.1), and Policy Identifiers) may have potential security or privacy impact if they are controlled by the requester. An attacker can leverage these extensions to create un-revocable certificates (by controlling the revocation endpoints - [example](https://randomoracle.wordpress.com/2009/07/31/ocsp-this-fail-brought-to-you-by-the-number-three/)), choose favorable CA chains (if multiple chains exist), observe client IP addresses (by placing their own CA certificate URL in the AIA extension and observing incoming requests), and so on.
2. **Dangerous domain-specific X.509 extensions**: If the same root CA is also used as a trust anchor for other types of certificates within an organization, and those scenarios rely on special extensions (e.g. [CanSignHttpsExchanges](https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#cross-origin-cert-req)) without requiring additional/special EKUs, then an attacker might be able to (ab)use Istio to issue such certificates, which they would then use to attack those non-Istio clients.

In short, it seems infeasible to validate every possible X.509 field or extension in a CSR. Given that the external CA chain may be trusted by a wider set of clients (not just Istio Agent) within an organization, and those clients may attach special significance to fields that appear in a certificate, it would be better to just avoid copying any unvalidated extensions and simply stick to a limited set of known extensions that the RA would validate.
